eradiate/scenes/biosphere/_discrete.py

```diff
@@ -909,6 +1028,159 @@ def bsdfs(self, ctx=None):
+        default="None",
+    )
+
+    trunk_height = documented(
+        pinttr.ib(default=0.0 * ureg.m, units=ucc.deferred("length")),
+        doc="Trunk height. If unset defaults to zero, yielding no trunk.\n\n"
+        "Unit-enabled field (default: ucc[length]).",
+        type="float",
+        default="0.0 m",
+    )
+
+    trunk_radius = documented(
+        pinttr.ib(default=0.0 * ureg.m, units=ucc.deferred("length")),
+        doc="Trunk radius. If unset defaults to zero, yielding no trunk.\n\nUnit-enabled field (default: ucc[length]).",
+        type="float",
+        default="0.0 m",
+    )
+
+    trunk_reflectance = documented(
+        attr.ib(
+            default=0.5,
+            converter=SpectrumFactory.converter("reflectance"),
+            validator=[
+                attr.validators.instance_of(Spectrum),
+                validators.has_quantity("reflectance"),
+            ],
+        ),
+        doc="Reflectance spectrum of the trunk. "
+        "Must be a reflectance spectrum (dimensionless).",
+        type=":class:`.Spectrum`",
+        default="0.5",
+    )
+
+    def bsdfs(self, ctx=None):
+        """
+        Return BSDF plugin specifications.
+
+        Parameter ``ctx`` (:class:`.KernelDictContext` or None):
+            A context data structure containing parameters relevant for kernel
+            dictionary generation.
+
+        Returns → dict:
+            Return a dictionary suitable for merge with a :class:`.KernelDict`
+            containing all the BSDFs attached to the shapes in the leaf cloud.
+        """
+
+        bsdfs_dict = self.leaf_cloud.bsdfs(ctx=ctx)
+
+        bsdfs_dict[f"bsdf_{self.id}"] = {
+            "type": "diffuse",
+            "reflectance": self.trunk_reflectance.kernel_dict(ctx=ctx)["spectrum"],
+        }
+
+        return bsdfs_dict
+
+    def shapes(self, ctx=None):
+        """
+        Return shape plugin specifications.
+
+        Parameter ``ctx`` (:class:`.KernelDictContext` or None):
+            A context data structure containing parameters relevant for kernel
+            dictionary generation.
+
+        Returns → dict:
+            A dictionary suitable for merge with a
+            :class:`~eradiate.scenes.core.KernelDict` containing all the shapes
+            in the canopy.
+        """
+        from mitsuba.core import ScalarTransform4f
+
+        kernel_length = uck.get("length")
+
+        kernel_height = self.trunk_height.m_as(kernel_length)
+        kernel_radius = self.trunk_radius.m_as(kernel_length)
+
+        # offset the leaf cloud by the size of the trunk
+        self.leaf_cloud.leaf_positions += (
+            np.array([0, 0, kernel_height]).reshape((1, 3)) * kernel_length
+        )
+
+        if ctx.ref:
+            bsdf = {"type": "ref", "id": f"bsdf_{self.id}"}
+        else:
+            bsdf = self.bsdfs(ctx=ctx)[f"bsdf_{self.id}"]
+
+        shapes_dict = self.leaf_cloud.shapes(ctx=ctx)
+
+        shapes_dict["trunk_cyl"] = {
```

Comment:
I think if you put multiple trees in your scene, this entry will be overwritten. Mind the shape naming pattern, see other classes for examples.